### ISSUE TITLE
Serever-side pagination to make ArgoCD UI a rocket.

### ISSUE THREAD

**Issue by Kulagin-G** [argoproj/argo-cd#14947]
# Summary

We're on ArgoCD `2.4.2+c6d0c8b` with more than 2000 Applications and suffering from an annoying issue with slow UI loading default `/applications` endpoint.  JFY: Gzip 9 is enabled on ArgoCD side and also on the Ingress level.

Depending on the user's internet connection and location UI loading can take 10s - 20s that's really slow and makes us think about the doubtful workaround, like redirecting users from `/application` endpoint to the custom one with a dummy Application. 

# Proposal

Endpoint `/application` returns one huge JSON that represents a set of ALL Applications managed by ArgoCD with all values, metadata, statuses, annotations, and the same for all managed k8s resources. It requires quite a lot of work on the client side and also on the server side for gzipping a big bunch of data.

Per discussion with @crenshaw-dev in the `#argo-cd` Slack channel, he mentioned that server-side pagination is one of the solutions that might be implemented to reduce data downloading on the client-side and increase UI performance significantly.

It would be great if the pagination will be customized, but it depends on implementation.

I guess it should be a really awaited feature due to there are a lot of issues/discussions about slow UI that were already reported/started.  Example - https://github.com/argoproj/argo-cd/discussions/8446 where @rumstead already collected related issues.

I hope this is possible and really appreciate your time. Thanks!

**Comment by alexymantha** [comment]
Server-side pagination was implemented by Akuity (see their blog post https://akuity.io/blog/akuity-july-2023-update/). When we asked on Slack, we were told that the team at Akuity was planning on upstreaming it in around a month (see discussion [here](https://cloud-native.slack.com/archives/C04SURUPDL2/p1691093380854169?thread_ts=1691087780.996729&cid=C04SURUPDL2).

**Comment by Whisper40** [comment]
That's great news, really! 
Currently with 4K applications, the UI (/applications) is clearly no longer usable.

**Comment by prashant-shahi** [comment]
@alexymantha Looking forward to the upstreaming  🎉

**Comment by crenshaw-dev** [comment]
@alexmt's original estimate would put this in time for 2.9, so setting it for that milestone for now. :-)

**Comment by Whisper40** [comment]
I really hope to see it backported for 2.7 release

**Comment by crenshaw-dev** [comment]
@Whisper40 I haven't seen the code, but my guess is that it was fairly involved. I kind of doubt we'll be able to backport, but will certainly consider it.

In good news, 2.8 seems to be pretty solid, and I expect 2.9 to be a relatively light release. So hopefully the upgrade path will be painless.

**Comment by Whisper40** [comment]
@crenshaw-dev @alexymantha Hi, it seems that the 2.9 release has been published but i don't see any mention about pagination. Is it normal ?

**Comment by taliastocks** [comment]
@Whisper40 I don't think this has been announced to be included in any particular release yet.

**Comment by jujubetsz** [comment]
It looks like it is going to be released in `v2.10` https://github.com/orgs/argoproj/projects/25/views/14

**Comment by crenshaw-dev** [comment]
The Akuity folks have told me they plan to get this into 2.10, so I've added the ticket to that column in the project. It's not a promise, just an indication of the current plan. :-)

**Comment by csantanapr** [comment]
Any progress on this from Akuity?

**Comment by dyatlov** [comment]
I'd say, if they won't merge their patch soon it can be someone else with a completely different implementation and then Akuity would have a hard time to maintain their fork.

**Comment by csantanapr** [comment]
@dyatlov are you signing up for the task?

**Comment by alexmt** [comment]
Sorry, for the delay with the PR. We ran into several issues along the way and had to take a step back to ensure good quality. As a short summary:

* CLI changes were absent. Akuity doesn’t provide customized Argo CD CLI to customers, so no related changes were implemented.
* Pagination is leveraging direct DB query to the etcd database. This is possible only if K8S is using KINE with DB as a backend. So this part has to be reworked as well.

Since the implementation is more involved than expected, I’ve started with a proposal that documents required changes: https://github.com/argoproj/argo-cd/pull/17222

**Comment by christophercorn** [comment]
Hello ArgoCD Team. What's the latest on when we think this feature will be released?

**Comment by ezxfv** [comment]
I found out that during the list application phase, we're getting back a bunch of data we don't really need, like `operationState` and `resources`. Since we're gonna fetch these again from the server once someone clicks on a specific application, how about we trim down the data during the list application phase? Could this also speed up the page response time?

**Comment by deanb-everc** [comment]
Hi guys, any update on this? would really help us a out.

**Comment by AnubhavSabharwa** [comment]
This is much awaited feature we are waiting for.
As we are maintaining 4000 apps and the UI becomes slow and sometimes it is dead slow.

**Comment by crenshaw-dev** [comment]
> i add a function to prune application contents, and it looks like working well, In my ArgoCD test environment (v2.9.0), the original API response returned a JSON with 15.5k lines and a size of 616KB. After modifications, the JSON now has 4.45k lines and a size of 128KB. This represents a 71.3% reduction in the number of lines and a 79.2% decrease in data volume.

@ezxfv to prioritize that change, I think we'd need reason to believe that payload is a significant performance bottleneck. For now, it seems like CPU is the more significant bottleneck.

**Comment by lucsoft** [comment]
@crenshaw-dev yeah TTFB (Time to first byte) should be a metric of interest.

**Comment by ezxfv** [comment]
@crenshaw-dev  Indeed, this method of reducing irrelevant information can only temporarily alleviate my current predicament, allowing the Argo CD homepage to barely load all applications (8000+). Upon a cursory examination of the performance on Chrome, the primary time consumers are script execution (7233ms) and rendering (4582ms). Would leveraging Web Workers and WASM contribute to accelerating this process? It appears that pagination remains the least affected solution by the growth in the number of applications.

**Comment by crenshaw-dev** [comment]
@ezxfv I imagine there are a lot of things that could be tuned on the frontend to speed things up. But even if you manage to half the script execution/rendering time, you're probably not going to get as much a win as you would be cutting the number of processed applications by 90%.

There are a lot of ways we could improve performance. Pagination seems like the highest payoff for the least effort / maintenance burden.

**Comment by rdelpret** [comment]
Any news on this?

**Comment by leoluz** [comment]
> Any news on this?

The proposal Alex wrote is now merged: https://github.com/argoproj/argo-cd/pull/17222

Implementation is pending.

**Comment by LeoAnt02** [comment]
Any update on this ? 
Maintaining a lot of apps (above 3-4k) start to feel painful with the UI
I think having argo cd ui running on things like Nextjs with a full server side first approach would be a good thing 
But this is probably bigger workload too

**Comment by ezxfv** [comment]
I made the following two small changes: by delaying the rendering of non-essential components, the entire page now loads much faster, and memory usage has been significantly reduced. I have 9500 applications, and now the homepage opens in just 1-2 seconds, with memory usage significantly reduced. Search, filtering, and other functions have become incredibly smooth. Previously, it took over 20 seconds, and the page could even crash.

![image](https://github.com/user-attachments/assets/359f5102-edf1-4b75-8592-65e66cf99819)

**Comment by tooptoop4** [comment]
@ezxfv where PR

**Comment by ezxfv** [comment]
> @ezxfv where PR
This is just a temporary version for review. If needed, I’ll organize it further later.

https://github.com/argoproj/argo-cd/pull/21334

**Comment by ezxfv** [comment]
> Any update on this ? Maintaining a lot of apps (above 3-4k) start to feel painful with the UI I think having argo cd ui running on things like Nextjs with a full server side first approach would be a good thing But this is probably bigger workload too

@LeoAnt02 Maybe you can build a new image with my pr, it supports 10w+ applications

**Comment by netashiloh** [comment]
@ezxfv Hi! In the company I work for we have an ArgoCD environment of over 3200 apps, and your solution sounds great to resolve the UI slowness.

We are now considering to split ArgoCD to multiple instances in order to solve the slowness issue. Will this solution be released in a new version?
If so, do you have an ETA?

**Comment by netashiloh** [comment]
@naturboy Our issue is mostly the first one you specified- the initial loading time of the webinterface of ArgoCD.
It actually improved for us after some server optimizations we did:

1. Increase the number of ARGOCD_API_SERVER_REPLICAS
2. Increase ARGOCD_GRPC_MAX_SIZE_MB value (we increased it to 1600).
3. Updated the resources limits and requests to limit: cpu 4, memory 12Gi and requests: cpu: 3, memory 8Gi

Note: Don't take our values as best practice, it's still a trial an error to see what's working for us. For reference, we work in a growing environment of 3000-7000 apps (we have dynamic envs).

There is still room for improvement and I'm sure the server side pagination will help. I plan to review [high availability argocd doc](https://argo-cd.readthedocs.io/en/stable/operator-manual/high_availability/) for more improvements we can do. Can you share the recommended configurations of RBAC too?

**Comment by ezxfv** [comment]
@netashiloh Split into multiple  instances doesn't resolve the ui problem, the problem  is many unnecessary  react dom node created in the background. I don't know whether my pr would be accepted or not. You can build  a image from my branch and test it.

**Comment by ezxfv** [comment]
i maybe add virtual page soon，when list more than 1000 applications at one time

**Comment by DanielYWoo** [comment]
We have 4000+ apps and the UI does not work now. Hopefully we can have this sooner

**Comment by ashish111333** [comment]
does @ezxfv  PR fixes this ?

**Comment by njsaeli** [comment]
Any update on this?

**Comment by masteryangel** [comment]
Hi!  Any update on this?  Thank you :)

**Comment by DanielYWoo** [comment]
seems there is a discussion in https://github.com/argoproj/argo-cd/pull/22444

**Comment by rumstead** [comment]
TL;DR:
Argo CD's UI struggles with performance when managing thousands of applications due to the /applications endpoint returning large, unpaginated payloads. A proposal to implement a paginated API endpoint was submitted in [PR #17222](https://github.com/argoproj/argo-cd/pull/17222) and approved.

The current active implementation effort is tracked in [PR #22444](https://github.com/argoproj/argo-cd/pull/22444), which is still open and under review. This PR aims to bring server-side pagination to the server-side components, and the UI and CLI may need additional changes. The PR needs reviewers.

**Comment by Leletir** [comment]
Any ETA on this feature ?

**Comment by todaywasawesome** [comment]
@Leletir No ETA as far as I can tell. The PR that was in draft seems to be dead. @gnunn1 said it might fit into a project they're working on. Both Codefresh and Akuity have solutions for this in their respective platforms so might be worth a conversation there aswell. 

I'd say contributions are open on this one and in the meantime RBAC can be leveraged to limit how many apps load for any individual user and that has been a solution for many orgs.

**Comment by crenshaw-dev** [comment]
One of my teammates has also started investigating the list UI from a frontend performance perspective. He suspects that we can make meaningful progress there.

**Comment by DanielYWoo** [comment]
> [@Leletir](https://github.com/Leletir) No ETA as far as I can tell. The PR that was in draft seems to be dead. [@gnunn1](https://github.com/gnunn1) said it might fit into a project they're working on. Both Codefresh and Akuity have solutions for this in their respective platforms so might be worth a conversation there aswell.
> 
> I'd say contributions are open on this one and in the meantime RBAC can be leveraged to limit how many apps load for any individual user and that has been a solution for many orgs.

A single microservice team can leverage RBAC to filter out irrelevant apps, but Ops, audit, productivity team, performance team, and security teams must see all apps. This is a critical requirement for large scale enterprise users where you have many different roles to access ArgoCD for full list of apps with server-side pagination.

**Comment by dhruvang1** [comment]
I have initiated a working PR (with small quirks) at https://github.com/argoproj/argo-cd/pull/25097

**Comment by jwinters01** [comment]
hey folks, I did some investigation into any optimizations we can make from the UI side to improve performance ahead of server side pagination, and made a quick PR #25201 to fix some unnecessary components that were rendering, which should be a decent performance improvements. I also have some additional ideas on how we can improve UI performance in a more macro sense that I outlined in a few issues I created:
#25203
#25204
These issues are more macro-level performance improvements but I think they will help the performance of chunky components like the Applications list in particular.

**Comment by DanielYWoo** [comment]
any update here? we have 5000+ apps :-)

**Comment by crenshaw-dev** [comment]
@DanielYWoo, @jwinters01's client-side improvements were merged and cherry-picked. They're available in 3.2.1. We're continuing to investigate client-side improvements, including upgrading React to a more recent version.

I wouldn't expect any night-and-day fixes soon, but expect incremental improvements over the next couple releases. :-)

**Comment by crenshaw-dev** [comment]
I just pushed 3.2.1 to an instance with ~1600 apps. Lighthouse score went 56->63 with "total blocking time" dropping from 450ms to 220ms. Again, not night-and-day, but the list UI should be somewhat snappier now.

**Comment by domdomegg** [comment]
For anyone hitting this with many apps — the `proj` query param on the applications list is applied server-side before serialization, so bookmarking a project-filtered URL is a significant speedup over the unfiltered homepage:

    https://<your-argocd>/applications?proj=team-a,team-b,team-c

**Comment by gadsilva** [comment]
Data point in support of this: we run a ~6,660-application instance where the Applications list takes ~2.5-4.3s server-side (~187 MB uncompressed payload) and has caused argocd-server OOMKills under concurrency  so server-side pagination/filtering would be a significant win for us.
There's a concrete implementation effort in #27786 (reportedly tested at 5k-20k apps); we're happy to help validate at scale on our instance.

**Comment by DanielYWoo** [comment]
> Data point in support of this: we run a ~6,660-application instance where the Applications list takes ~2.5-4.3s server-side (~187 MB uncompressed payload) and has caused argocd-server OOMKills under concurrency so server-side pagination/filtering would be a significant win for us. There's a concrete implementation effort in [#27786](https://github.com/argoproj/argo-cd/issues/27786) (reportedly tested at 5k-20k apps); we're happy to help validate at scale on our instance.

We  have roughly the same number of apps and the homepage (/applications) never loads. :-)
So we built another tool to index, search and give the directly app link which works.

**Comment by gadsilva** [comment]
@DanielYWoo nice, curious about your tool:

Is it open-source / would you be open to sharing it?
How do you keep the index fresh (poll the ArgoCD API, watch, or an external source of truth)?
Do you link straight to /applications/<ns>/<app>, and did you have to handle SSO/deep-link auth?

**Comment by DanielYWoo** [comment]
not that fancy as you expected ;-)
we have a CMDB where tracks the microservice to argoCD app mappings, to open an app in ArgoCD, you go to the CMDB and click the link. This is definitely a workaround, not a standalone opensource software purposed for improving ArgoCD.

**Comment by gadsilva** [comment]
> not that fancy as you expected ;-) we have a CMDB where tracks the microservice to argoCD app mappings, to open an app in ArgoCD, you go to the CMDB and click the link. This is definitely a workaround, not a standalone opensource software purposed for improving ArgoCD.

I understand. Thank you so much for the feedback.

**Comment by crenshaw-dev** [comment]
afaik there's no fully functional open PR. Server-side pagination is incredibly complex and past attempts have fizzled.

We've just completed an upgrade to React 19, which brings inherent performant improvements and significantly improved profiling capabilities, so we can start identifying and resolving low-hanging client-side performance issues.

There's also an [open PR](https://github.com/argoproj/argo-cd/pull/25451) to improve server-side field filtering for the applications list endpoint. That PR will reduce network latency by significantly reducing the payload size.

If we exhaust payload and client-side optimizations without resolving performance for large-scale instances, I think it'll be time to return to pagination.

**Comment by crenshaw-dev** [comment]
I'm merging the payload improvement for 3.5, so we'll get a small boost with that release, especially for apps that manage lots of resources. But I suspect network won't be the main bottleneck for many users. The React work will follow.  :-)

**Comment by RyeNoodles** [comment]
👋 We updated Argo CD from `v3.4.1` to `v3.5.0-rc1` and noticed a massive improvement in UI loading speed - great work and thanks to everyone involved! Unfortunately we had to roll back as we hit a bug where `ServerSideApply` is effectively ignored and large CRDs fail to sync. Same app worked fine on v3.4.1 (quite literally, we rolled back the version and we could sync it):

```
CustomResourceDefinition.apiextensions.k8s.io "something.io" is invalid: metadata.annotations: Too long: may not be more than 262144 bytes
```

It fails in the dry-run phase even with `ServerSideApply=true`. Root cause _looks like_ (this is pure LLM pre-research since I'm not familiar with the codebase) the new dry-run block in `gitops-engine` `pkg/sync/sync_context.go` (`applyObject`) for existing non-scheme resources (CRDs) it now sets `dryRunStrategy = DryRunServer`, but `shouldUseServerSideApply` still returns `false` for any dry-run. So the dry-run runs as client-side apply + server dry-run, which sends the oversized `last-applied-configuration` annotation to the API server and gets rejected. In `v3.4.1` the dry-run was `DryRunClient`, which validated locally and passed, then the real apply went server-side.

It fails in the dry-run phase even with `ServerSideApply=true`. Reproducible standalone with any CRD >256 KiB:

```
kubectl apply --dry-run=server -f bigcrd.yaml                                 # fails: annotations Too long
kubectl apply --server-side --force-conflicts --dry-run=server -f bigcrd.yaml # succeeds
```

Looking forward to use v3.5.0 soon again 🤞

**Comment by crenshaw-dev** [comment]
Glad to hear about the speedup!

Can you open a separate issue for the SSA bug?
